### PR TITLE
[Storage] Add fast GCS listFrom using the startOffset list parameter

### LINKED ISSUE #7456: [Storage] GCSLogStore.listFrom lists the entire _delta_log per call — port the S3 startAfter optimization to GCS startOffset
### Which Delta project/connector is this regarding?
- [x] Other (delta-storage / GCSLogStore)

### Overview

`GCSLogStore` inherits `HadoopFileSystemLogStore.listFrom`, which does
`fs.listStatus(parent)` — a full directory listing, O(directory size), on **every**
log metadata operation. S3 got an opt-in fix for exactly this in #1191 (flag
`delta.enableFastS3AListFrom`, pushes `startAfter` into the S3 LIST request). GCS has
an equivalent server-side primitive — the `startOffset` parameter of objects.list —
but nothing uses it.

### Motivation (production numbers, GCS, Aug 2026)

Measured on a production lakehouse (Delta tables on GCS, asia-southeast1):

- A `_delta_log` with 337K objects cost ~17s per `listFrom` (47 sequential list pages
  through gcs-connector). A pipeline that touches Delta metadata per event spent
  **93.7% of its wall clock** inside these listings; trimming the directory gave an
  immediate 6× end-to-end speedup — without touching compute.
- Directories this large are easy to reach: log cleanup deletes expired `<v>.json`
  but never `<v>.crc` version-checksum files (they accumulate forever), and tables
  with multi-part checkpoints retain `checkpoint parts × retained versions` objects.
  Our worst table's `_delta_log` holds **1.7M objects** (2,343-part checkpoints);
  one listFrom there costs ~90s. With startOffset pushdown it returns in well under
  a second (only keys ≥ the requested version are ever transferred).

### Further details

GCS's `startOffset` is *inclusive* (S3's `startAfter` is exclusive), so the port is
simpler than the S3 version — no "key just before" computation is needed.

Proposed shape, mirroring the S3 implementation:
- `io.delta.storage.internal.GCSLogStoreUtil`: lists via
  `com.google.cloud:google-cloud-storage`'s
  `Storage.list(bucket, prefix(dir), startOffset(resolvedKey), currentDirectory(), fields(NAME, SIZE, UPDATED))`,
  converts blobs to `FileStatus`.
- `GCSLogStore.listFrom` override gated by a new Hadoop conf flag
  **`delta.enableFastGCSListFrom`** (default **false**, matching the S3 convention),
  skipping Local/RawLocal filesystems, with an `fs.exists(parent)` probe to preserve
  listFrom's documented FileNotFoundException contract (prefix-listing a missing
  "directory" returns empty rather than throwing).
- `google-cloud-storage` added to delta-storage as **provided** (compile-only), pinned
  to 2.61.0 — the version gcs-connector 4.0.4 (the repo's current pin) resolves via
  libraries-bom 26.73.0. The class is only loaded when the flag is enabled.

Notes for reviewers:
- The gcs-connector cannot supply this natively today: gcsio's `ListObjectOptions`
  exposes no startOffset, and the shaded connector relocates its embedded java-storage
  client (`com.google.cloud.hadoop.repackaged.gcs.*`), so its client/credentials are
  not reusable from delta-storage. The util therefore authenticates via Application
  Default Credentials; `fs.gs.auth.*` keyfile configs are not honored (documented
  limitation of the opt-in flag).

### Willingness to contribute

- [x] Yes. I have an implementation with unit tests and production benchmark numbers
  and will open a PR.

### PR DESCRIPTION
**Pull request by AadhiKat**
#### Which Delta project/connector is this regarding?
- [x] Other (Storage)

## Description

Resolves #7456

`GCSLogStore` inherits `HadoopFileSystemLogStore.listFrom`, which lists the **entire** parent
directory (`fs.listStatus(parent)`) and filters client-side — O(directory size) on every log
metadata operation. S3 received an opt-in fix for exactly this in #1191
(`delta.enableFastS3AListFrom`, pushing `startAfter` into the LIST request). GCS has the
equivalent server-side primitive — the `startOffset` parameter of `objects.list` — and since
gcs-connector 4.0.4 (the version this repo pins, #7109) the connector exposes it directly on
the Hadoop entry class: `GoogleHadoopFileSystem.listStatusStartingFrom(Path)`
(hadoop-connectors#1551, backported to 4.0.x in hadoop-connectors#1760).

Large `_delta_log` directories are easy to accumulate in practice: metadata cleanup deletes
expired `<version>.json` but never `<version>.crc` version-checksum files, and tables with
multi-part checkpoints retain `parts × retained checkpoints` objects. On one production table
whose `_delta_log` holds **3,231,409 objects** (2,428-part checkpoints), the default `listFrom`
is effectively unusable: the raw list operation alone takes ~173–175 s (per the connector's
own `op_list_status` instrumentation) and then materializing the directory OOMs both 4 GB and
12 GB heaps — while the same `listFrom` result (21 files) comes back in ~1 s through this
PR's fast path on a default 4 GB heap.

This PR adds a `listFrom` override in `GCSLogStore`, gated by a new Hadoop conf flag
**`delta.enableFastGCSListFrom`** (default **false**, matching the S3 convention), built
entirely on the connector's own API:

- **Same identity as every other LogStore operation.** The fast path lists through the same
  `GoogleHadoopFileSystem` instance used by `read`/`write`/`exists` — same credentials, same
  configuration, per-Hadoop-conf rather than JVM-wide (mirroring how the S3 fast path stays on
  `S3AFileSystem`). If the `gs://` FileSystem is not a `GoogleHadoopFileSystem`, or the
  connector predates `listStatusStartingFrom` (a `LinkageError` is caught once and remembered),
  the default implementation is used automatically.
- **No new dependency.** The previous revision's `com.google.cloud:google-cloud-storage`
  dependency is gone; delta-storage compiles against the `shaded`-classifier gcs-connector the
  repo already pins (provided scope, mirroring `hadoop-aws` for the S3 path). The entry class
  and method are not relocated in the shaded jar, and their signatures only use Hadoop types,
  so nothing unshaded is required on any classpath.
- **Lazy, memory-bounded listing.** `GCSLogStoreUtil.gcsListFrom` returns an iterator that
  holds at most one list page (`fs.gs.list.max.items.per.call`, default 5000 entries) and
  fetches the next page only when the consumer crosses the boundary, resuming server-side from
  the NUL successor of the last listed key (`startOffset` is inclusive). Consumers that stop
  early issue no further requests.
- `listFrom`'s documented `FileNotFoundException` contract is preserved with an explicit
  `fs.exists(parent)` probe (a prefix listing of a missing directory returns empty rather than
  throwing), structurally identical to `S3SingleDriverLogStore`.

## How was this patch tested?

- New unit tests, all running in regular CI against a faithful in-memory
  `GoogleHadoopFileSystem` stub:
  - `GCSLogStoreUtilTest`: server-side offset arguments (inclusive start, NUL-successor
    resume), lazy page fetching, termination on leaving the parent directory, directory
    placeholders at page boundaries, subdirectory-entry skipping, eligibility/unwrap logic,
    bucket-root and non-connector guards.
  - `GCSLogStoreFastListFromSuite`: missing parent → `FileNotFoundException`; past latest
    version → empty iterator; flag off / local fs → default implementation; fast-vs-default
    parity on one fixture; `LinkageError` → sticky fallback to the default implementation.
- `GCSLogStoreUtilIntegrationTest` (env-gated) now constructs `GCSLogStore` with the flag ON
  and exercises `listFrom` end-to-end against real GCS through the connector, with
  `fs.gs.list.max.items.per.call=2` so pagination and resume are exercised, asserting the
  exact `startOffset` values passed to the connector (server-side narrowing), plus
  FNFE/past-latest/subdirectory/parity/metadata cases.
- `startOffset` semantics (inclusive; NUL successor) validated against the GCS JSON API
  directly.
- Measured on production GCS `_delta_log`s (cross-region, median of 3, identical results
  fast-vs-default in every completed comparison):
  - 85,481-object log: default **13.7 s** vs fast **1.05 s** (~13x), same 9 entries.
  - 3,231,409-object log (2,428-part checkpoints): fast **~1.0 s** for the 21-file tail on a
    default 4 GB heap; the default path's raw list op alone takes **~173-175 s** and then
    OOMs materializing the directory (reproduced at 4 GB and 12 GB heaps).
- Reproducible synthetic fixture (84,001-object `_delta_log`: 40k jsons + 40k crcs +
  20x200-part checkpoints, generated by a 10-line script): listing a 2,202-entry tail,
  default **~11.9 s** vs fast **~0.9 s** (~13x), identical entries across all runs.

## Does this PR introduce _any_ user-facing changes?

No behavior change by default. With `delta.enableFastGCSListFrom=true` (new opt-in Hadoop
conf), `GCSLogStore.listFrom` returns the same files faster. Requires gcs-connector 4.0.4+
(`GoogleHadoopFileSystem.listStatusStartingFrom`); with older connectors or non-connector
`gs://` FileSystems the flag is ignored and the default implementation is used. Note: like the
S3 fast path, the fast listing returns only files (no directory entries for subdirectories of
`_delta_log`).

### CODE REVIEW

File: storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+     * the gcs-connector may use: the connector's shaded client cannot be reused here because
+     * its google-cloud-storage classes are relocated.
+     */
+    private static volatile Storage storage;
+
+    private static Storage getStorage() {
+        if (storage == null) {
+            synchronized (GCSLogStoreUtil.class) {
+                if (storage == null) {
+                    storage = StorageOptions.getDefaultInstance().getService();
```
Comment (murali-db) [inline storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java]: this lists via `storageoptions.getdefaultinstance()`, so it's adc, not the hadoop/gcs-connector identity that `read`/`write`/`exists` use.

with the flag on, `listfrom` can fail while the fs still works, or succeed as a different principal and return a different set than a later `read`.

s3's fast path stays on `s3afilesystem` so creds stay consistent.

could we build the storage client from the same hadoop conf (`google.cloud.auth.service.account.json.keyfile` / `fs.gs.auth.*`), or fall back to `super.listfrom` if we can't reproduce that identity?

also the client is a jvm-wide singleton, so one identity for every gcs table in the process.

File: build.sbt [hunk 2]
```diff
@@ -1275,6 +1275,10 @@ lazy val storage = (project in file("storage"))
       // Note that the org.apache.hadoop.fs.s3a.Listing::createFileStatusListingIterator 3.3.1 API
       // is not compatible with 3.3.2.
       "org.apache.hadoop" % "hadoop-aws" % hadoopVersion % "provided",
+
+      // Only needed at runtime when delta.enableFastGCSListFrom is enabled (GCSLogStoreUtil).
+      // Version aligned with what gcs-connector 4.0.4 pulls in (via libraries-bom 26.73.0).
+      "com.google.cloud" % "google-cloud-storage" % "2.61.0" % "provided",
```
Comment (murali-db) [inline build.sbt]: this adds unshaded `google-cloud-storage` 2.61.0 as provided.

we shade gcs-connector so unshaded google libs (protobuf 4) don't collide with spark/hadoop protobuf 3. see #7109.

the connector's copy is relocated, so it won't satisfy these imports.

enabling the flag as written can `nosuchmethoderror` the spark jobs this is meant to help.

could we list with hadoop creds without putting the unshaded sdk on the user classpath (json `objects.list`, or shade it inside delta-storage)?

if we keep the sdk, we should document a spark + shaded-connector recipe that actually works, and test the flag on in that setup.

File: storage/src/main/java/io/delta/storage/GCSLogStore.java [hunk 3]
```diff
@@ -60,10 +66,50 @@ public class GCSLogStore extends HadoopFileSystemLogStore {
...
+     */
+    private final boolean enableFastListFrom
+            = initHadoopConf().getBoolean("delta.enableFastGCSListFrom", false);
+
     public GCSLogStore(Configuration hadoopConf) {
         super(hadoopConf);
     }
 
+    @Override
+    public Iterator<FileStatus> listFrom(Path path, Configuration hadoopConf) throws IOException {
```
Comment (murali-db) [inline storage/src/main/java/io/delta/storage/GCSLogStore.java:92]: the interesting bits of this override never run in ci.

`publicgcslogstoresuite` is on local fs so it always takes the slow path, and the integration test calls `gcslogstoreutil` directly so it never hits `exists()` -> `filenotfoundexception` vs empty-past-latest.

could we add a fake filesystem unit test for: missing parent -> fnfe, past latest version -> empty iterator, flag off / local fs -> `super.listfrom`?

and if we have a real gcs fixture, construct `gcslogstore` with the flag on rather than only calling the util.

File: storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                Storage.BlobListOption.fields(
+                    Storage.BlobField.NAME, Storage.BlobField.SIZE, Storage.BlobField.UPDATED)
+            ).iterateAll();
+        } catch (com.google.cloud.storage.StorageException e) {
+            throw new IOException(
+                String.format("Failed to list gs://%s/%s", bucket, dirKey), e);
+        }
+        final List<FileStatus> statuses = new ArrayList<>();
+        for (Blob blob : blobs) {
+            statuses.add(blobToFileStatus(blob, parentPath));
```
Comment (huan233usc) [inline storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java]: It seems that we collect iterator to the list and then make it an iterator?
Also if #file is large, any memory concern?

File: storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java (hunk 1 above)
Comment (AadhiKat) [reply storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java]: Reworked it to stay on the filesystem like the S3 path does: gcs-connector 4.0.4 (what this repo pins after #7109) actually ships `GoogleHadoopFileSystem.listStatusStartingFrom(Path)` (added in hadoop-connectors#1551, backported to 4.0.x in #1760). It's not relocated in the shaded jar and the signature is all hadoop types, so delta-storage can call it directly. Since it goes through the fs instance, the listing runs with exactly the creds/config of the other LogStore ops. No client of our own anymore, so the JVM-wide singleton problem is gone too.

If the fs isn't a GoogleHadoopFileSystem, or the connector is older than the API (LinkageError), listFrom falls back to super.listFrom.

File: build.sbt (hunk 2 above)
Comment (AadhiKat) [reply build.sbt]: Removed the SDK dependency completely. delta-storage now compiles against the shaded-classifier gcs-connector already pinned in this repo (provided scope, same as hadoop-aws for the S3 path). I checked the published 4.0.4 shaded jar - the entry class and listStatusStartingFrom aren't relocated, so nothing unshaded is needed at runtime either. The flag works on a stock shaded-connector deployment, and the integration suite now runs flag-on on exactly that setup.

Re-ran the numbers on this revision against a production _delta_log with 3,231,409 objects (cross-region): the fast path returns the 21-file tail in ~1s on the default 4G heap. The default implementation didn't finish at all - the raw list op takes ~175s (connector's own op_list_status metric) and then it OOMs building the FileStatus array. Tried 4G and 12G heaps, dies on both.

File: storage/src/main/java/io/delta/storage/GCSLogStore.java (hunk 3 above)
Comment (AadhiKat) [reply storage/src/main/java/io/delta/storage/GCSLogStore.java:92]: Added those three as unit tests (missing parent -> FNFE, past latest version -> empty iterator, flag off / local fs -> super.listFrom), plus pagination resume, prefix truncation, subdirectory skipping and the LinkageError fallback. They run against an in-memory GoogleHadoopFileSystem stub that mimics the connector's list semantics (flat bucket-wide scan, raw page cut before placeholder filtering), so all of it runs in normal CI.

The integration test now goes through GCSLogStore with the flag on instead of calling the util directly, covers the FNFE and past-latest cases against real GCS, and sets fs.gs.list.max.items.per.call=2 so pagination actually kicks in with a handful of files. It also records the startOffset of every request to check the narrowing really happens server-side.

File: storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java (hunk 4 above)
Comment (AadhiKat) [reply storage/src/main/java/io/delta/storage/internal/GCSLogStoreUtil.java]: Changed it: listFrom now returns an iterator that pages lazily through listStatusStartingFrom, so at most one list page (fs.gs.list.max.items.per.call, default 5000 entries) is in memory at a time, and callers that stop early don't fetch the remaining pages at all. (The S3 util has the same TODO about this.)

The default path is where this really bites btw - on a _delta_log with 3.2M objects, the fs.listStatus based listFrom OOM'd at 4G and at 12G in my tests. The lazy path handles the same listing in ~1s on 4G.

**Comment by AadhiKat** [comment]
Hey @murali-db @huan233usc @raveeram-db @sumeet-db @tdas - Can you please have a look? I've addressed the comments and ran the updated test suites.
